File: backend/internal/playback/playback_test.go
package playback

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// fixtureCase mirrors one entry of backend/testdata/playback_cases.json, the
// file shared with the frontend test-suite.
type fixtureCase struct {
	Name         string `json:"name"`
	Items        []Item `json:"items"`
	AnchorMillis int64  `json:"anchorMillis"`
	NowMillis    int64  `json:"nowMillis"`
	Expect       *struct {
		Index                 int   `json:"index"`
		PlaylistItemID        int64 `json:"playlistItemId"`
		MediaID               int64 `json:"mediaId"`
		CycleIndex            int64 `json:"cycleIndex"`
		ElapsedInCycleMillis  int64 `json:"elapsedInCycleMillis"`
		LoopIteration         int64 `json:"loopIteration"`
		ElapsedInItemMillis   int64 `json:"elapsedInItemMillis"`
		RemainingInItemMillis int64 `json:"remainingInItemMillis"`
		TruncatedByCycle      bool  `json:"truncatedByCycle"`
	} `json:"expect"`
}

type fixtureFile struct {
	CycleMillis int64         `json:"cycleMillis"`
	Cases       []fixtureCase `json:"cases"`
}

func loadFixtures(t *testing.T) fixtureFile {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "..", "testdata", "playback_cases.json"))
	if err != nil {
		t.Fatalf("read fixtures: %v", err)
	}
	var f fixtureFile
	if err := json.Unmarshal(raw, &f); err != nil {
		t.Fatalf("parse fixtures: %v", err)
	}
	if len(f.Cases) == 0 {
		t.Fatal("fixture file contains no cases")
	}
	return f
}

func TestCycleMillisMatchesFixture(t *testing.T) {
	f := loadFixtures(t)
	if f.CycleMillis != CycleMillis {
		t.Fatalf("cycle length mismatch: fixture %d, code %d", f.CycleMillis, CycleMillis)
	}
	if CycleMillis != 18_000_000 {
		t.Fatalf("the assignment fixes the cycle at 5 hours (18000s), got %dms", CycleMillis)
	}
}

func TestResolveAgainstSharedFixtures(t *testing.T) {
	f := loadFixtures(t)
	for _, tc := range f.Cases {
		t.Run(tc.Name, func(t *testing.T) {
			state, ok := Resolve(NewTimeline(tc.Items), tc.AnchorMillis, tc.NowMillis)

			if tc.Expect == nil {
				if ok {
					t.Fatalf("expected no playable item, got index %d", state.Index)
				}
				return
			}
			if !ok {
				t.Fatal("expected a playable item, got none")
			}

			checks := []struct {
				field string
				got   int64
				want  int64
			}{
				{"index", int64(state.Index), int64(tc.Expect.Index)},
				{"playlistItemId", state.Item.PlaylistItemID, tc.Expect.PlaylistItemID},
				{"mediaId", state.Item.MediaID, tc.Expect.MediaID},
				{"cycleIndex", state.CycleIndex, tc.Expect.CycleIndex},
				{"elapsedInCycleMillis", state.ElapsedInCycleMillis, tc.Expect.ElapsedInCycleMillis},
				{"loopIteration", state.LoopIteration, tc.Expect.LoopIteration},
				{"elapsedInItemMillis", state.ElapsedInItemMillis, tc.Expect.ElapsedInItemMillis},
				{"remainingInItemMillis", state.RemainingInItemMillis, tc.Expect.RemainingInItemMillis},
			}
			for _, c := range checks {
				if c.got != c.want {
					t.Errorf("%s = %d, want %d", c.field, c.got, c.want)
				}
			}
			if state.TruncatedByCycle != tc.Expect.TruncatedByCycle {
				t.Errorf("truncatedByCycle = %v, want %v", state.TruncatedByCycle, tc.Expect.TruncatedByCycle)
			}
		})
	}
}

// TestPlaybackIsContinuous walks a whole 5-hour cycle item by item and asserts
// there is never a gap, an overlap, or a stall between playlist items.
func TestPlaybackIsContinuous(t *testing.T) {
	items := []Item{
		{PlaylistItemID: 1, MediaID: 101, DurationMillis: 10_000},
		{PlaylistItemID: 2, MediaID: 102, DurationMillis: 20_000},
		{PlaylistItemID: 3, MediaID: 103, DurationMillis: 30_000},
	}
	tl := NewTimeline(items)

	var now int64
	transitions := 0
	for now < CycleMillis {
		state, ok := Resolve(tl, 0, now)
		if !ok {
			t.Fatalf("playback stopped at %dms", now)
		}
		if state.RemainingInItemMillis <= 0 {
			t.Fatalf("non-advancing state at %dms", now)
		}
		// The instant the current item ends, the next item must already own it.
		next := now + state.RemainingInItemMillis
		if next < CycleMillis {
			following, ok := Resolve(tl, 0, next)
			if !ok {
				t.Fatalf("playback stopped at handover %dms", next)
			}
			if following.ElapsedInItemMillis != 0 {
				t.Fatalf("gap at %dms: next item started %dms in", next, following.ElapsedInItemMillis)
			}
			wantIndex := (state.Index + 1) % tl.Len()
			if following.Index != wantIndex {
				t.Fatalf("at %dms expected item index %d, got %d", next, wantIndex, following.Index)
			}
		}
		now = next
		transitions++
	}
	// 18000s / 60s per pass * 3 items.
	if transitions != 900 {
		t.Fatalf("expected 900 item transitions in one cycle, got %d", transitions)
	}
}

// TestCycleBoundaryNeverBlanks asserts the remaining cycle time is filled by
// looping the playlist rather than by blank playback.
func TestCycleBoundaryNeverBlanks(t *testing.T) {
	// 7s playlist: 18000000 / 7000 = 2571 whole passes with 3000ms left over.
	tl := NewTimeline([]Item{{PlaylistItemID: 7, MediaID: 701, DurationMillis: 7_000}})

	for _, now := range []int64{17_997_000, 17_998_500, 17_999_999} {
		state, ok := Resolve(tl, 0, now)
		if !ok {
			t.Fatalf("no item at %dms - the tail of the cycle went blank", now)
		}
		if state.Item.MediaID != 701 {
			t.Fatalf("unexpected media %d at %dms", state.Item.MediaID, now)
		}
	}

	// And the very next millisecond re-anchors to the top of the playlist.
	state, ok := Resolve(tl, 0, CycleMillis)
	if !ok {
		t.Fatal("no item at the cycle boundary")
	}
	if state.CycleIndex != 1 || state.ElapsedInItemMillis != 0 || state.LoopIteration != 0 {
		t.Fatalf("cycle did not re-anchor cleanly: %+v", state)
	}
}

func TestTimelineSkipsUnplayableItems(t *testing.T) {
	tl := NewTimeline([]Item{
		{PlaylistItemID: 1, MediaID: 101, DurationMillis: 5_000},
		{PlaylistItemID: 2, MediaID: 102, DurationMillis: 0},
		{PlaylistItemID: 3, MediaID: 103, DurationMillis: -1},
	})
	if tl.Len() != 1 {
		t.Fatalf("expected 1 playable item, got %d", tl.Len())
	}
	if tl.TotalMillis() != 5_000 {
		t.Fatalf("expected total 5000ms, got %d", tl.TotalMillis())
	}
}

func TestFloorDiv(t *testing.T) {
	cases := []struct{ a, b, want int64 }{
		{0, 10, 0},
		{9, 10, 0},
		{10, 10, 1},
		{-1, 10, -1},
		{-10, 10, -1},
		{-11, 10, -2},
	}
	for _, c := range cases {
		if got := floorDiv(c.a, c.b); got != c.want {
			t.Errorf("floorDiv(%d, %d) = %d, want %d", c.a, c.b, got, c.want)
		}
	}
}

File: frontend/src/utils/playback.test.js
import { describe, expect, it } from 'vitest';

import fixtures from '../../../backend/testdata/playback_cases.json';
import {
  CYCLE_MILLIS,
  buildTimeline,
  occurrenceKey,
  resolvePlayback,
  timelineFromPlaylist,
} from './playback.js';

describe('shared fixtures', () => {
  it('agrees with the backend on the cycle length', () => {
    expect(CYCLE_MILLIS).toBe(fixtures.cycleMillis);
    expect(CYCLE_MILLIS).toBe(18_000_000);
  });

  // Every case in this file is also asserted by the Go test-suite, which is
  // what guarantees the two playback clocks stay identical.
  for (const testCase of fixtures.cases) {
    it(testCase.name, () => {
      const state = resolvePlayback(
        buildTimeline(testCase.items),
        testCase.anchorMillis,
        testCase.nowMillis,
      );

      if (testCase.expect === null) {
        expect(state).toBeNull();
        return;
      }

      expect(state).not.toBeNull();
      expect(state.index).toBe(testCase.expect.index);
      expect(state.item.playlistItemId).toBe(testCase.expect.playlistItemId);
      expect(state.item.mediaId).toBe(testCase.expect.mediaId);
      expect(state.cycleIndex).toBe(testCase.expect.cycleIndex);
      expect(state.elapsedInCycleMillis).toBe(testCase.expect.elapsedInCycleMillis);
      expect(state.loopIteration).toBe(testCase.expect.loopIteration);
      expect(state.elapsedInItemMillis).toBe(testCase.expect.elapsedInItemMillis);
      expect(state.remainingInItemMillis).toBe(testCase.expect.remainingInItemMillis);
      expect(state.truncatedByCycle).toBe(testCase.expect.truncatedByCycle);
    });
  }
});

describe('continuous playback', () => {
  const playlist = buildTimeline([
    { playlistItemId: 1, mediaId: 101, durationMillis: 10_000 },
    { playlistItemId: 2, mediaId: 102, durationMillis: 20_000 },
    { playlistItemId: 3, mediaId: 103, durationMillis: 30_000 },
  ]);

  it('hands over from one item to the next with no gap and no stall', () => {
    let now = 0;
    let transitions = 0;

    while (now < CYCLE_MILLIS) {
      const state = resolvePlayback(playlist, 0, now);
      expect(state).not.toBeNull();
      expect(state.remainingInItemMillis).toBeGreaterThan(0);

      const next = now + state.remainingInItemMillis;
      if (next < CYCLE_MILLIS) {
        const following = resolvePlayback(playlist, 0, next);
        // The instant an item ends, the next one already owns the clock.
        expect(following.elapsedInItemMillis).toBe(0);
        expect(following.index).toBe((state.index + 1) % playlist.items.length);
      }
      now = next;
      transitions += 1;
    }
    // 18000s / 60s per pass * 3 items.
    expect(transitions).toBe(900);
  });

  it('loops the playlist rather than blanking the rest of the cycle', () => {
    // A 7s playlist divides the cycle unevenly: 2571 passes plus 3000ms.
    const short = buildTimeline([{ playlistItemId: 7, mediaId: 701, durationMillis: 7_000 }]);

    for (const now of [17_997_000, 17_998_500, 17_999_999]) {
      const state = resolvePlayback(short, 0, now);
      expect(state).not.toBeNull();
      expect(state.item.mediaId).toBe(701);
    }
  });

  it('re-anchors to playlist position 0 at the 5 hour boundary', () => {
    const last = resolvePlayback(playlist, 0, CYCLE_MILLIS - 1);
    expect(last.cycleIndex).toBe(0);
    expect(last.loopIteration).toBe(299);

    const first = resolvePlayback(playlist, 0, CYCLE_MILLIS);
    expect(first.cycleIndex).toBe(1);
    expect(first.index).toBe(0);
    expect(first.elapsedInItemMillis).toBe(0);
    expect(first.loopIteration).toBe(0);
  });

  it('resolves the same item regardless of when a client joins', () => {
    // Two clients computing at the same instant must agree; this is what makes
    // a page refresh seamless.
    const instant = 1_234_567;
    expect(resolvePlayback(playlist, 0, instant)).toEqual(resolvePlayback(playlist, 0, instant));
  });
});

describe('timelineFromPlaylist', () => {
  const apiPlaylist = [
    { id: 11, position: 0, media: { id: 1, name: 'M1', type: 'image', durationSeconds: 10 } },
    { id: 12, position: 1, media: { id: 2, name: 'M7', type: 'blank', durationSeconds: 5 } },
    { id: 13, position: 2, media: { id: 3, name: 'M3', type: 'video', durationSeconds: 30 } },
  ];

  it('converts API playlists into the millisecond timeline', () => {
    const timeline = timelineFromPlaylist(apiPlaylist);
    expect(timeline.totalMillis).toBe(45_000);
    expect(timeline.starts).toEqual([0, 10_000, 15_000]);
  });

  it('treats a configured blank item as ordinary playable media', () => {
    const timeline = timelineFromPlaylist(apiPlaylist);
    const state = resolvePlayback(timeline, 0, 12_000);
    expect(state.item.mediaId).toBe(2);
    expect(state.remainingInItemMillis).toBe(3_000);
  });

  it('returns no state for an empty or missing playlist', () => {
    expect(resolvePlayback(timelineFromPlaylist([]), 0, 5_000)).toBeNull();
    expect(resolvePlayback(timelineFromPlaylist(undefined), 0, 5_000)).toBeNull();
  });
});

describe('occurrenceKey', () => {
  const timeline = buildTimeline([
    { playlistItemId: 1, mediaId: 101, durationMillis: 10_000 },
    { playlistItemId: 2, mediaId: 102, durationMillis: 20_000 },
  ]);

  it('is stable while one item plays', () => {
    const a = occurrenceKey(resolvePlayback(timeline, 0, 1_000));
    const b = occurrenceKey(resolvePlayback(timeline, 0, 9_999));
    expect(a).toBe(b);
  });

  it('changes when the item changes and when the playlist comes round again', () => {
    const first = occurrenceKey(resolvePlayback(timeline, 0, 1_000));
    const second = occurrenceKey(resolvePlayback(timeline, 0, 11_000));
    const firstAgain = occurrenceKey(resolvePlayback(timeline, 0, 31_000));

    expect(second).not.toBe(first);
    // Same item, next pass: the key must change so a video restarts.
    expect(firstAgain).not.toBe(first);
  });

  it('handles the empty state', () => {
    expect(occurrenceKey(null)).toBe('empty');
  });
});
